### PR TITLE
ICS721 v2: cw721 v19 with creator and onchain extensioms (aka metadata)

### PR DESCRIPTION
**Pull request by taitruong**
This is ics721 v2 using cw721 v19 (includes collection extension (metadata) and creator):
- instantiates cw721 v19 on initial NFT transfer
  - transfers optional collection extension(e.g. description, royalties, ...)
  - sets creator, using cw721 admin
- SG specifics
  - collection data is retrieved from sg721
  - also sg721 is instantiated on initial transfer
- tranfers optional NFT extensions and mints NFT with extension (onchain metadata)
- tests covered:
  - source NFT can be of version v15, v16, v17, v18, and v19
  - escrowed, instantiated collection is ofc always v19
- other
  - updated libs

### CODE REVIEW

**Review by shanev — APPROVED** [review]

File: packages/ics721/src/utils.rs [hunk 1]
```diff
@@ -1,52 +1,72 @@
...
+        name,
+        symbol,
+        extension,
+        updated_at: _,
+    } = deps.querier.query_wasm_smart(
         collection,
-        &cw721_base::msg::QueryMsg::<Empty>::ContractInfo {},
+        #[allow(deprecated)]
+        // For now we use `ContractInfo` which is known across all version, whilst `GetCollectionInfoAndExtension` is only available in v0.19 and higher
+        &cw721_metadata_onchain::msg::QueryMsg::ContractInfo {},
```
Comment (taitruong) [inline packages/ics721/src/utils.rs:46]: This is where collection metadata is retrieved. cw721 v18 and below only holds name and symbol. Whilst v19 holds creator and royalty info.

File: contracts/sg-ics721/src/execute.rs [hunk 2]
```diff
@@ -16,28 +20,52 @@ impl Ics721Execute for SgIcs721Contract {
...
+            share: r.share,
+        });
+        let extension = Some(CollectionExtension {
+            description: collection_info.description,
+            image: collection_info.image,
+            external_link: collection_info.external_link,
+            explicit_content: collection_info.explicit_content,
+            start_trading_time: collection_info.start_trading_time,
+            royalty_info,
+        });
```
Comment (taitruong) [inline contracts/sg-ics721/src/execute.rs:40]: Here sg-ics721 overrides default implementation and gets collection data from sg721-base via `CollectionInfo` query.

File: packages/ics721/src/execute.rs [hunk 3]
```diff
@@ -623,6 +643,21 @@ where
...
+                        image: Some(ext.image),
+                        external_link: ext.external_link,
+                        explicit_content: ext.explicit_content,
+                        start_trading_time: ext.start_trading_time,
+                        royalty_info: ext.royalty_info.map(|r| RoyaltyInfoResponse {
+                            payment_address: cw721_admin_or_ics721_admin_or_ics721_creator, // r.payment_address cant be used, since it is from another chain
+                            share: r.share,
+                        }),
+                    });
+            instantiate_msg.collection_info_extension = collection_info_extension_msg;
```
Comment (taitruong) [inline packages/ics721/src/execute.rs:660]: on ibc receive for initial transfer, collection extension is passed to cw721 instantiate message.

File: packages/ics721/src/execute.rs [hunk 4]
```diff
@@ -410,12 +414,17 @@ where
...
         // in this case ICS721 may have metadata stored
-        let token_metadata = TOKEN_METADATA
+        let token_metadata = match IBC_RECEIVE_TOKEN_METADATA
             .may_load(deps.storage, (class.id.clone(), token_id.clone()))?
-            .flatten();
+            .flatten()
+        {
+            Some(metadata) => Some(metadata),
+            // incase there is none in the storage, this is the 'home' chain, so metadata is retrieved from the cw721 contract
+            None => info.extension.map(|ext| to_json_binary(&ext)).transpose()?,
```
Comment (taitruong) [inline packages/ics721/src/execute.rs:426]: Here, on source chain, nft extension (metadata) is retrieved and passed in ibc packet.

File: packages/ics721/src/execute.rs [hunk 5]
```diff
@@ -684,13 +721,35 @@ where
...
+                            background_color: ext.background_color,
+                            description: ext.description,
+                            external_url: ext.external_url,
+                            image: ext.image,
+                            image_data: ext.image_data,
+                            youtube_url: ext.youtube_url,
+                            name: ext.name,
+                        }),
+                    None => None,
+                };
```
Comment (taitruong) [inline packages/ics721/src/execute.rs:746]: On receive, ics721 checks whether ibc packet holds nft extensions - if so NFT is minted with onchain metadata.

File: packages/ics721/src/utils.rs [hunk 6]
```diff
@@ -1,52 +1,72 @@
...
+                    // cw721 v0.16 and lower holds minter (simple string storage)
+                    let minter_response: cw721_base_016::msg::MinterResponse =
+                        deps.querier.query_wasm_smart(
+                            collection,
+                            &cw721_base_016::QueryMsg::Minter::<Empty> {},
+                        )?;
+                    deps.api.addr_validate(&minter_response.minter)?;
+                    Some(minter_response.minter)
+                }
+            }
```
Comment (taitruong) [inline packages/ics721/src/utils.rs:33]: pls also note various versions checks here for cw721 v15 and higher.

File: ts-relayer-tests/src/ics721.spec.ts [hunk 7]
```diff
@@ -377,6 +441,43 @@ test.serial("transfer NFT: wasmd -> osmo", async (t) => {
...
+  if (wasmCollectionData.extension?.royalty_info?.payment_address) {
+    wasmCollectionData.extension.royalty_info.payment_address = osmoAddr; // osmo cw721 admin address is used as payment address
+  }
+  t.deepEqual(
+    wasmCollectionData,
+    osmoCollectionData,
+    `collection data must match: ${JSON.stringify(
+      wasmCollectionData
+    )} vs ${JSON.stringify(osmoCollectionData)}`
+  );
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:463]: integration check whether collection extension on target chain matches with source chain

File: ts-relayer-tests/src/ics721.spec.ts [hunk 8]
```diff
@@ -377,6 +441,43 @@ test.serial("transfer NFT: wasmd -> osmo", async (t) => {
...
+  const minterOwnerShip = await getCw721MinterOwnership(osmoClient, osmoCw721);
+  t.is(osmoIcs721, minterOwnerShip.owner);
+
+  const wasmNftInfo = await nftInfo(wasmClient, wasmCw721, tokenId);
+  // assert extension is not null
+  t.truthy(wasmNftInfo.extension);
+  const osmoNftInfo = await nftInfo(osmoClient, osmoCw721, tokenId);
+  t.truthy(osmoNftInfo.extension);
+  // assert nft with extension is same
+  t.deepEqual(wasmNftInfo, osmoNftInfo);
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:480]: check for nft onchain metadata

File: ts-relayer-tests/src/ics721.spec.ts [hunk 9]
```diff
@@ -746,6 +847,471 @@ test.serial("transfer NFT: wasmd -> osmo", async (t) => {
   t.is(wasmAddr, tokenOwner.owner);
 });
 
+test.serial("transfer NFT v16: wasm -> osmo", async (t) => {
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:850]: same test, but with cw721 v16 on source chain

File: ts-relayer-tests/src/ics721.spec.ts [hunk 10]
```diff
@@ -746,6 +847,471 @@ test.serial("transfer NFT: wasmd -> osmo", async (t) => {
...
+  );
+  const osmoCollectionData = await getCw721CollectionInfoAndExtension(
+    osmoClient,
+    osmoCw721
+  );
+  t.deepEqual(
+    wasmCollectionData.name,
+    osmoCollectionData.name,
+    `name must match: ${wasmCollectionData.name} vs ${osmoCollectionData.name}`
+  );
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:929]: cw721 only holds name and symbole in collection extension

File: ts-relayer-tests/src/ics721.spec.ts [hunk 11]
```diff
@@ -746,6 +847,471 @@ test.serial("transfer NFT: wasmd -> osmo", async (t) => {
...
+  const minterOwnerShip = await getCw721Minter_v16(osmoClient, osmoCw721);
+  t.is(osmoIcs721, minterOwnerShip.minter);
+
+  const wasmNftInfo = await nftInfo(wasmClient, wasmCw721_v16, tokenId);
+  // assert extension is not null
+  t.truthy(wasmNftInfo.extension);
+  const osmoNftInfo = await nftInfo(osmoClient, osmoCw721, tokenId);
+  t.truthy(osmoNftInfo.extension);
+  // assert nft with extension is same
+  t.deepEqual(wasmNftInfo, osmoNftInfo);
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:945]: cw721 v16 and higher holds same nft extension (as defined by erc721 spec).

**Review by humanalgorithm — APPROVED** [review]

**Review by jhernandezb — APPROVED** [review]

**Pull request merged**